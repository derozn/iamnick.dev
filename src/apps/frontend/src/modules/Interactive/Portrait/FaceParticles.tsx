import { useRef, useEffect } from 'react';
import { useLoader, useFrame, extend } from '@react-three/fiber';
import { TextureLoader } from 'three';

import { ParticlesMaterial } from './materials/particles';
import { Particles } from './objects/particles';

extend({ ParticlesMaterial, Particles });

export const FaceParticles = () => {
  const particles = useRef<Particles>(null);
  const particlesMaterialRef = useRef<ParticlesMaterial>(null);
  const particleTexture = useLoader(TextureLoader, '/textures/nick.png');

  useEffect(() => {
    if (particles.current !== null) particles.current.create();
  }, []);

  useFrame((_, delta) => {
    if (particlesMaterialRef.current) {
      particlesMaterialRef.current.update(delta);
    }
  });

  return (
    <group name="particle-hit-group">
      <mesh name="particles" position={[75, 0, 0]}>
        <particles args={[particleTexture]} attach="geometry" ref={particles} />
        <particlesMaterial
          ref={particlesMaterialRef}
          attach="material"
          particleTexture={particleTexture}
          extensions-derivatives
        />
      </mesh>
    </group>
  );
};

export default FaceParticles;
