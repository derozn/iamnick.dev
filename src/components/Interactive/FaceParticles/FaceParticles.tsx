import React, { FunctionComponent, useRef, useEffect } from 'react';
import { useLoader, useFrame, extend } from 'react-three-fiber';
import { TextureLoader, Vector3 } from 'three';

import { ParticlesMaterial } from './materials/particles';
import { Particles } from './objects/particles';

extend({ ParticlesMaterial, Particles });

type Props = {
  textureUrl: string;
  position?: number[];
};

const FaceParticles: FunctionComponent<Props> = ({ textureUrl, position }) => {
  const particles = useRef<Particles>(null);
  const particlesMaterialRef = useRef<ParticlesMaterial>(null);
  const particleTexture = useLoader(TextureLoader, textureUrl);

  useEffect(() => {
    if (particles.current !== null) particles.current.create();
  }, []);

  useFrame((_, delta) => {
    if (particlesMaterialRef.current) {
      particlesMaterialRef.current.update(delta);
    }
  });

  return (
    <group name="particle-hit-group">
      <mesh name="particles" position={(position as unknown) as Vector3}>
        <particles args={[particleTexture]} attach="geometry" ref={particles} />
        <particlesMaterial
          ref={particlesMaterialRef}
          attach="material"
          particleTexture={particleTexture}
          extensions-derivatives
        />
      </mesh>
    </group>
  );
};

export default FaceParticles;
