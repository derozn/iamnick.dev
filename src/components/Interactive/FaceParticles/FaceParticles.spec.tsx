import React, { Suspense } from 'react';
import { create } from 'react-test-renderer';
import { useLoader } from 'react-three-fiber';
import { TextureLoader } from 'three';

import useStore from '#store';
import FaceParticles from './FaceParticles';

jest.mock('#store');
jest.mock('react-three-fiber');

const mockUseLoader = useLoader as jest.Mock<any>;
const mockUseStore = useStore as jest.Mock<any>;

describe('components/FaceParticles', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders FaceParticles', () => {
    mockUseStore.mockImplementation((cb) => {
      return cb({ face: { textureUrl: 'texture-url' } });
    });

    mockUseLoader.mockImplementationOnce(() => 'texture');

    const wrapper = create(
      <Suspense fallback={null}>
        <FaceParticles />
      </Suspense>,
    );

    expect(mockUseLoader).toHaveBeenCalledWith(TextureLoader, 'texture-url');
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
