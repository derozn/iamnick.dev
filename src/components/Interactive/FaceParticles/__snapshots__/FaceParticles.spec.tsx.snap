// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FaceParticles renders FaceParticles 1`] = `
<group
  name="particle-hit-group"
>
  <mesh
    name="particles"
  >
    <instancedBufferGeometry
      attach="geometry"
    />
    <faceMaterial
      attach="material"
      extensions-derivatives={true}
      faceTexture="texture"
      time={0}
    />
  </mesh>
</group>
`;
